### ISSUE TITLE
RAML 1.0 support - progress, questions, help needed, etc

### ISSUE THREAD

**Issue by kevinrenskers** [raml2html/raml2html#254]
I've just published raml2obj 4.0.0-beta1 with the most basic RAML 1.0 support, and raml2html 4.0.0-beta1 which uses that new version of raml2obj. However, **the templates in raml2html do not work with RAML 1.0 files, so we don't have RAML 1.0 support yet**!

There have been 2 branches here to add RAML 1.0 support, but in my opinion we should not continue with those. Adding RAML 1.0 support in the current templates is a waste of time since we're now going to build new templates anyway. I also think we really need to focus on making raml2obj spit out a unified format across RAML 0.8 and 1.0 files. That way clients like raml2html and raml2md only have to worry about rendering that format, instead of having a bunch of if-statements and hacks to support both RAML versions. raml2obj needs to do the heavy lifting.

So..
1. Remove the two RAML 1.0 branches and focus on the develop branch instead.
2. Make raml2obj do the heavy lifting and spit out a unified format.
3. Write new templates using that new unified format.

**What I need from the RAML Workgroup, or someone from the community, is awesome example files for both RAML 0.8 and 1.0 which include all the features that raml2html needs to support. I can then see how raml2obj needs to change.**

I do believe that this is the most sensible way forward, but it does mean we have to focus on raml2obj, RAML 1.0 support and unified JS object, before even starting on the new templates.

**Comment by kevinrenskers** [comment]
I've gone ahead and removed the old raml1.0 branch, but the raml1.0_new branch is still there. All in favor of removing it and focusing on making raml2obj spit out a unified file format? 

Is there actually any talk about the official RAML parser doing this? Would make sense that *it* would give us a unified format to make everyones life easier.

**Comment by andjarnic** [comment]
So, is RAML2HTML fully 1.0 supported now? If not I would argue that merging into master at this time wouldnt make sense until either raml2obj supported a unified output or the parser did.

**Comment by kevinrenskers** [comment]
"However, the templates in raml2html do not work with RAML 1.0 files, so we don't have RAML 1.0 support yet!"

Nobody is talking about merging anything to master, not sure what you mean?

**Comment by sichvoge** [comment]
> What I need from the RAML Workgroup, or someone from the community, is awesome example files for both RAML 0.8 and 1.0 which include all the features that raml2html needs to support. 

Two repositories that I can "sponsor" :)

- https://github.com/raml-org/raml-examples
- https://github.com/raml-apis

**Comment by entertainyou** [comment]
@kevinrenskers , it's there ETA for raml 1.0 support? Is there something community can help?

**Comment by kevinrenskers** [comment]
RAML 1.0 support plus new template should be done before September 30. The first task is to get raml2obj to output a consistent object no matter then RAML version used (0.8 or 1.0). I'd be very happy with community help, especially once a beta is done to get people to test it, etc.

**Comment by fesor** [comment]
> I'd be very happy with community help

I see only one problem with compatibility of 0.8 and 1.0 output format, this is schemas. All other stuff could be easily converted from 0.8 to 1.0 format. We could probably generate type declarations from json schemas for consistency and this will simplify things a lot. What do you think?

**Comment by kevinduffey** [comment]
Annotations... what is being done to add support for some defined annotations. For example, I mentioned this some time back, it would be a great idea for the RAML community to put out a set of annotations that all parsers should try to support, such as deprecation, versioning of resources, name spacing, etc. I think annotations are a powerful concept but it requires every parser to know what to parse for and how to use the annotation and it would go a long way in the RAML community coming up with a few common/requested annotations and support them in RAML2HTML among other projects. Thoughts?

**Comment by kevinrenskers** [comment]
@fesor sounds good, but I just don't know enough about it yet :)
@kevinduffey if it's not in the standard then we shouldn't add it.

**Comment by andjarnic** [comment]
@kevinrenskers I have to disagree to some extent. I understand RAML2HTML aims to support RAML 1.0. But the spec adds the ability to handle annotations. It unfortunately requires each and every RAML parsing project to add support for specific annotations to make it viable. It is intended to allow customization to the spec, however, as RAML2HTML is the go to API doc generator for RAML, I think it behooves the RAML work group and community to have the defacto HTML API doc generator support and render some commonly needed/requested/*should be in the spec* annotations. Being able to render something like the use of deprecation for a given resource is rather easy, and extremely useful. Unless I am missing something there isnt a way to really do this at this time without something like the use of annotations and RAML2HTML rendering that as part of the process. 

I dont think this has to be in the initial RAML2HTML 1.0 work.. but it is something I would rather see in the one doc generator rather than 4 or 5 other generators popping up to support their own thoughts of annotations. Then you end up with various tools that do similar things and fragmentation. That isnt to say RAML2HTML is the end all API doc generator, other people will customize templates, even build their own RAML parser to HTML generator.

My point is that RAML2HTML is one of the most often used projects (if not the most) for RAML, and its as good as any place to start the inclusion and rendering of some annotations that the spec itself may leave out, but REST APIs in general would benefit from and/or should have, such as resource deprecation, versioning details, maybe even HATEOAS details.

**Comment by andjarnic** [comment]
@kevinrenskers I have to disagree to some extent. I understand RAML2HTML aims to support RAML 1.0. But the spec adds the ability to handle annotations. It unfortunately requires each and every RAML parsing project to add support for specific annotations to make it viable. It is intended to allow customization to the spec, however, as RAML2HTML is the go to API doc generator for RAML, I think it behooves the RAML work group and community to have the defacto HTML API doc generator support and render some commonly needed/requested/*should be in the spec* annotations. Being able to render something like the use of deprecation for a given resource is rather easy, and extremely useful. Unless I am missing something there isnt a way to really do this at this time without something like the use of annotations and RAML2HTML rendering that as part of the process. 

I dont think this has to be in the initial RAML2HTML 1.0 work.. but it is something I would rather see in the one doc generator rather than 4 or 5 other generators popping up to support their own thoughts of annotations. Then you end up with various tools that do similar things and fragmentation. That isnt to say RAML2HTML is the end all API doc generator, other people will customize templates, even build their own RAML parser to HTML generator.

My point is that RAML2HTML is one of the most often used projects (if not the most) for RAML, and its as good as any place to start the inclusion and rendering of some annotations that the spec itself may leave out, but REST APIs in general would benefit from and/or should have, such as resource deprecation, versioning details, maybe even HATEOAS details.

**Comment by kevinrenskers** [comment]
You have some compelling arguments :) For me personally, I don't work on raml2html full time; I am back to work on RAML 1 support and new templates, and that's kind of about it. 

I simply don't use raml2html myself anymore and so have no need for those features. I'm afraid that this annotation support will have to come from someone else. I'm also not sure if it's correct for raml2html to push one annotations spec - like you said, this *should be in the spec* :)

**Comment by andjarnic** [comment]
Thanks. :D  My thought is that the RAML Workgroup/community come up with some needed/wished for annotations and we put together a Library for them, and then a separate Which projects support these.. sort of list. @sichvoge what do you think? Can we get this started? Let me know how I can help if so? I think at the very least we get a discussion going around some useful annotations that would make designing APIs with RAML better, more capable, etc.

**Comment by fesor** [comment]
@kevinrenskers I could prepare PR to make raml2obj generate consistent type info for both 0.8 and 1.0 but I will have some time on this only in the end of next week.

**Comment by sichvoge** [comment]
@fesor @kevinrenskers `schemas` and `types` are synonymous and therefore should work perfectly fine with either. No need for converting JSON schemas into RAML data types IMHO

**Comment by sichvoge** [comment]
@andjarnic why not creating a wish list and put them into an issue on the raml-spec github repo?

**Comment by kevinrenskers** [comment]
Hey guys, a quick update. I've started to work on raml2obj: to make its output consistent no matter if a RAML 0.8 or 1.0 source is used (and to add unit tests for this). A consistent output will make it a lot easier for raml2html and raml2md (and other projects using raml2obj) to support both RAML 0.8 and 1.0.

HOWEVER. I was thinking maybe we just drop 0.8 support. It will make the code a lot easier, less hacky, way more maintainable going forward. If you're not ready to move to 1.0 yet, you can always stick with raml2html 3.0.0. But moving your existing 0.8 files to 1.0 should be quite straightforward and simple, so there isn't really any reason to stick with 0.8.

So I think (and @sichvoge agrees) that dropping 0.8 support and only officially supporting 1.0 going forward makes a lot of sense. Simpler templates, simpler raml2obj, no real downsides. We can focus on 100% coverage for everything 1.0 supports without worrying about 0.8 compatibility.

Good idea? Huge mistake? Let me know :)

**Comment by sichvoge** [comment]
I agree to Kevin's concerns about maintainability and hacky code trying to support both versions. People will still be able to generate HTML from 0.8 since you have tags and you can easily install 3.0.0 which still has support for 0.8 from NPM. So its not that 0.8 completely vanished. But for the sake of simplicity, I would also only support 1.0.

**Comment by v-kolesnikov** [comment]
Good idea! 👍

**Comment by SinisterMinister** [comment]
So, my concern right now is the [RAML Projects Page](http://raml.org/projects/projects) lists this as currently supporting 1.0. Until 1.0 support is in beta, shouldn't that not be the case? It's very misleading for devs such as myself who expected this tool to be fully ready only to find out that it is, in fact, not.

**Comment by kevinduffey** [comment]
Agreed @SinisterMinister. At the very least projects that are in progress but not full baked, should indicate some status like Alpha, Beta, etc.

**Comment by sichvoge** [comment]
I updated the RAML projects page and removed the label for now.

**Comment by kevinrenskers** [comment]
I just released raml2obj 4.0.0-beta2. This has proper support for RAML 1, has a bunch of unit tests, and changes the output slightly to make it a lot easier to work with (no more nested objects wrapped in other objects inside of an array, etc). The output is not compatible with existing raml2html templates, but since we're going to ship brand new templates anyway, that's no problem (and existing users with RAML 0.8 files should just stay on raml2html 3.x).

**Comment by kevinrenskers** [comment]
I've just released raml2html 4.0.0-beta2. This is the first initial release with real RAML 1 support! It's still using the old templates, which have been fixed to use the slightly changed output from raml2obj. 

A lot is still to do of course, not in the least create the new templates, but it's exciting to have a solid foundation now.

I have removed the `raml-jsonschema-expander` stuff since I think it belongs in raml2obj instead. I have not added it there yet though, so the schema expanding isn't available at the moment. There are already enough spinning plates in the air, so I don't want to focus on this bit *just yet* :)

**Comment by kevinduffey** [comment]
I cant wait to dig in. Great work guys. So does this convert RAML Types to Java POJOs? I ask because there is some work on the RAML2JAXRS going on that is working on a stand alone tool to convert RAML to Java POJOs similar to the JSONSchema2Pojo library. It would be ideal if you are doing the same thing here in any way, that you guys all collaborate and try to utilize the one library, as opposed to different ways of doing the same thing. If you are not, no worries, just wasnt sure everyone here might be aware of that work as well.

**Comment by kevinrenskers** [comment]
No. The RAML parser gives us strings back (see also https://github.com/raml-org/raml-js-parser-2/issues/479), which for raml2html works kinda great for displaying them in an easy way. If in the future we want real objects so we can syntax highlight the types (for example), then we'd need to look into transforming them, and yup, in that case I'd prefer to use an existing project to do that :)

**Comment by kevinduffey** [comment]
Gah..my bad.. I forget this is a nodejs project too.. which voids the need for a java library. Wasnt thinking! I would suspect then for Atom/RAML IDE to already have some JS capability to parse the RAML Types since it can validate and such in the IDE?

**Comment by kevinrenskers** [comment]
Oh haha I kinda read over the "java" part of your question. My brain just read it as "javascript". 
Not sure what you mean with your second question, I don't think it has anything to do with raml2html?

**Comment by kevinrenskers** [comment]
Published raml2html 4.0.0-beta4. Still using the old templates, because that way I can just focus on supporting the new and changes RAML features instead of also having to worry about a new look :)

I disabled the `type` rendering for now, since it has become a confusing property. Sometimes it's just something like `string` or `number` while other times it's a fully fledged object schema. They can't be rendered in the same way of course, but they're all simply strings so I have to figure out how to make a difference between them. Maybe check for `{` in the type?

Anyway, basic support for annotations is added (on the method level only in the current templates), baseUriParameters are rendered, the `examples` array is supported, `properties` are now rendered. Also types from libraries are now expanded and rendered in place.

It's getting there, but daaaamn RAML sure is a complicated spec. I could really use help with testing this release. Please check what features are broken or missing, let's discuss in separate tickets. And then of course pull requests would be very welcome.

**Comment by sichvoge** [comment]
There was some initial implementation for the type support in one of the very old branches. Maybe that helps?!

**Comment by kevinrenskers** [comment]
It checked for `type.properties` to determine how to use the type. But that seems a bit strange to me, not every type necessarily has properties right? It could be `uriParameters` for example?

**Comment by sichvoge** [comment]
Don't remember how we did it in the raml1.0 branch a month ago, but it was working perfectly. I'd suggest to start there.

**Comment by kevinrenskers** [comment]
Yup I am looking at https://github.com/raml2html/raml2html/pull/248/files

**Comment by cody-greene** [comment]
I noticed the `lib/` directory is no longer included in the npm package b/c package.json has:
```json
"files": [
  "index.js",
  "bin/raml2html"
]
```

which means this doesn't work (because the default template is missing):
```
$ npm install -g raml2html@4.0.0-beta4
$ raml2html api.raml >index.html
Error: template not found: ./lib/template.nunjucks
  at createTemplate ...
```

**Comment by kevinrenskers** [comment]
Whoops! I'll fix this right away, thanks!

**Comment by kevinrenskers** [comment]
Alrighty, beta5 is published.

**Comment by kevinrenskers** [comment]
Beta 6 was published today, fixing some bugs. Please keep the input coming :)

**Comment by iwex** [comment]
Hello! Also could you please print all properties of object (if it's inherited from another) including parents.

**Comment by kevinrenskers** [comment]
Good point, hadn't really thought of that. Can you open a new issue in the raml2obj repo for that?

**Comment by ddidier** [comment]
Hi,
I'm using the beta-7 and found an issue when using (use:) shared data types at different levels. I attached 3 tests. Basically `test.raml` includes `dataTypes/Object1.raml` which includes `dataTypes/Object2.raml`. If both include (`use:`) `dataTypes/commons.raml` the following error is raised:

```
TypeError: Cannot read property 'indexOf' of null
    at AccumulatingRegistry.get (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:222:30)
    at AccumulatingRegistry.TypeRegistry.get (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/typesystem.js:307:33)
    at contributeToAccumulatingRegistry (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:845:50)
    at parse (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:731:18)
    at parsePropertyBean (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:372:19)
    at /usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:799:21
    at Array.forEach (native)
    at parse (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:797:34)
    at Object.parse (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/parse.js:726:27)
    at Object.parseTypeFromAST (/usr/local/lib/node_modules/raml2html/node_modules/raml-typesystem/dist/src/index.js:101:15)
``` 

[API-test-failing.zip](https://github.com/raml2html/raml2html/files/520203/API-test-failing.zip)
[API-test-passing-1.zip](https://github.com/raml2html/raml2html/files/520204/API-test-passing-1.zip)
[API-test-passing-2.zip](https://github.com/raml2html/raml2html/files/520205/API-test-passing-2.zip)

**Comment by kevinrenskers** [comment]
@sichvoge looks like an error in `raml-1-parser` to me? We're not using `raml-typesystem` directly in raml2html or raml2obj.

**Comment by sichvoge** [comment]
Let me take a look at that. thanks!

**Comment by kevinrenskers** [comment]
Beta 8 was published just now, with better support for type inheritance. There is still an issue with array types that I'm looking into, but I think that's a bug in the `datatype-expansion` library.

It also uses a newer version of `raml-1-parser`, so hopefully that error from @ddidier is also solved with that?

**Comment by sichvoge** [comment]
Did you raise the issue?

**Comment by kevinrenskers** [comment]
The array types in datatype-expansion? Yup.

**Comment by ddidier** [comment]
@kevinrenskers sorry but this isn't working either.

**test.raml**
```yaml
#%RAML 1.0

title: Test
version: v1
mediaType: application/json
/test:
  get:
    body:
      type: !include dataTypes/Object1.raml
```

**dataTypes/Object1.raml**
```yaml
#%RAML 1.0 DataType

type: object
properties:
  o1a: string
```

**raml2html test.raml**
```
{ Template render error: (/usr/local/lib/node_modules/raml2html/lib/template.nunjucks)
  Template render error: (/usr/local/lib/node_modules/raml2html/lib/template.nunjucks)
  Template render error: (/usr/local/lib/node_modules/raml2html/lib/resource.nunjucks) [Line 166, Column 44]
  TypeError: type.indexOf is not a function
    at Object.exports.prettifyError (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/lib.js:34:15)
    at /usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:486:31
    at new_cls.root [as rootRenderFunc] (eval at _compile (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:565:24), <anonymous>:191:3)
    at new_cls.render (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:479:15)
    at /usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:312:35
    at createTemplate (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:234:25)
    at handle (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:249:25)
    at /usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/environment.js:263:21
    at next (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/lib.js:207:13)
    at Object.exports.asyncIter (/usr/local/lib/node_modules/raml2html/node_modules/nunjucks/src/lib.js:214:5) name: 'Template render error' }
```

**Comment by kevinrenskers** [comment]
Thanks, I see the problem. It's because I expected all types to always be a string, but this one is an object. Sigh, RAML is way too complex ;)

Luckily the fix was very easy, beta 9 is published now.

**Comment by ddidier** [comment]
thank you, that's working

**Comment by ddidier** [comment]
I understand it's a work in progress but how can I access/print the whole objects structure. The default template only renders  `[object Object]`. I looked around a little in the source code but that's confusing... thank you.

**Comment by kevinrenskers** [comment]
At https://github.com/raml2html/raml2html/blob/develop/index.js#L26 you could console.log `ramlObj`, or use the `dump` filter in the template. It would be better to ask questions like these in a new issue though, rather then making this ticket even longer with unrelated questions :)

**Comment by ulaBLUE** [comment]
Hi,
is there any beta version for raml2md that supports RAML 1.0. Thanks

**Comment by kevinrenskers** [comment]
Yes, 4.0.0-beta1. See also https://github.com/raml2html/raml2md/tree/develop. Please keep questions about raml2md in its own repo, thanks!

**Comment by kevinrenskers** [comment]
Just a quick update: the RAML Workgroup is going to look into the datatype-expansion issue, hopefully that'll be fixed soon. I believe it's the biggest blocker to releasing raml2obj, raml2html and raml2md, so with a bit of luck it'll be released soon.

In the meantime you can just use the beta versions from NPM of course, and feedback is still very welcome! Preferably in new issues, not as comments in this ticket :)

**Comment by kevinrenskers** [comment]
I've published 4.0.0-beta11 which should solve a bunch of problems with JSON datatypes (see also [this](https://github.com/raml-org/raml-parser-toolbelt/issues/40#issuecomment-257542925)). I think we're about good to go and publish 4.0.0 soon.

**Comment by kevinrenskers** [comment]
And already beta12 is out as well, greatly improving array type rendering.

**Comment by SinisterMinister** [comment]
Just for clarification:

> I've published 4.0.0-beta11 which should solve a bunch of problems with JSON datatypes

This isn't referring to the datatype-expansion issue, right? Or perhaps I'm just misunderstanding how that is supposed to work. I'm expecting to see all of the parameters of an object no matter how nested it is, where right now (beta12) I'm seeing only the extended properties of the type:

![Type ss](https://cloud.githubusercontent.com/assets/800441/19894441/b80d32f4-a012-11e6-98af-1782380839ff.png)

Sorry if I'm just being oblivious. I blame the lack of coffee after a long weekend.

**Comment by jfstephe** [comment]
@kevinrenskers 
Hi,
I've been caught out by this today. I'm using the beta version now but I was trying out the '!include' notation for resourceTypes etc and AFAIK this isn't working (I can raise the appropriate defect if you would like - let me know).

My question, however, is what are the pending 1.0 features to fully support RAML v1.0, and when do you think they will be complete by?

Many thanks for your continuing efforts on this!

John

**Comment by kevinrenskers** [comment]
We have example RAML files that use the include notation so that surprises me. Best to open a new issue with an example file that fails. Thanks!

**Comment by kevinrenskers** [comment]
Oh yeah and to answer your question; well it's bugs like these that are stopping a final 4.0.0 release :) Let's get them ironed out and then we can finally release this.

**Comment by jfstephe** [comment]
:-). Ok I was wrong. I was (and after some tweaks am still) using https://github.com/walling/gulp-raml2html and that library sent me down a wrong path. After looking more closely at my raml files I spotted the problem and confidence is coming back ;-). Thanks for your quick response and sorry for the false alarm!

**Comment by kevinrenskers** [comment]
4.0.0 has been released! We know it's not 100% perfect yet, but the RAML 1.0 support is pretty solid, and it should be a good .0 release. It'll also get more people to use it, report problems, and hopefully contribute to the project.

**Comment by sichvoge** [comment]
Thanks Kevin for all that work. Now, I really hope that overs will hope on and contribute going the last mile.

**Comment by rlunadev** [comment]
![image](https://user-images.githubusercontent.com/7585253/38552235-7eb0a35a-3c89-11e8-9fe6-4838ab4c8503.png)
 for what?
